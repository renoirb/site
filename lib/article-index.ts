import { extractDateTuple } from './models/article'

export type FilterTypeFindByDate = (year: number, month?: number) => string[]

export const extractYears = (items: string[]): number[] => {
  const years = new Set<number>()
  for (const line of items) {
    const dateTuple = extractDateTuple(line)
    years.add(dateTuple[0])
  }

  return [...years].sort()
}

export const extractMonths = (items: string[]) => {
  const list = createFindByDate(items)
  return (year: number): number[] => {
    const months = new Set<number>()
    const subList = list(year)
    for (const line of subList) {
      const dateTuple = extractDateTuple(line)
      months.add(dateTuple[1])
    }

    return [...months].sort()
  }
}

export const createFindByDate = (items: string[]): FilterTypeFindByDate => (
  year: number = 0,
  month: number = 0
): string[] => {
  const lines = [...items].filter(line => {
    const dateTuple = extractDateTuple(line)
    if (dateTuple[0] === year) {
      if (month > 0) {
        if (dateTuple[1] === month) {
          return true
        } else {
          return false
        }
      }
      return true
    }
    return false
  })
  return lines
}

export class ArticleIndex {
  private lines: string[] = []

  /**
   * True if we have collection of lines that contains a path separator.
   *
   * If at the first add, the line had a path separator,
   * any other add will only work if they also contain a path separator.
   */
  private hasPaths: boolean = false

  /**
   * True if we have a collection of lines where the first characters contains a date.
   *
   * Only format that toggles this to true is if a line starts by a YYYY-MM-DD format
   * e.g. 2019-05-22-Foo-bar.
   *
   * Otherwise set to false.
   */
  private threeFirstAreNumbers: boolean = false

  constructor(readonly indexFile: string = '') {}

  setLines(lines: string[]): void {
    lines.forEach(line => this.add(line))
  }

  add(line: string): void {
    const isFirstAdd = this.lines.length < 1
    const hasPaths = /\//.test(line)
    const threeFirstAreNumbers = extractDateTuple(line) !== false
    if (isFirstAdd) {
      this.hasPaths = hasPaths
      this.threeFirstAreNumbers = threeFirstAreNumbers
    } else {
      let message = `Lines can only be with one format possible, we expected the line `
      if (hasPaths !== this.hasPaths) {
        if (this.hasPaths) {
          message += 'TO CONTAIN a path separator (e.g. /).'
        } else {
          message += 'NOT to contain a path separator.'
        }
        throw new Error(message)
      }
      if (threeFirstAreNumbers !== this.threeFirstAreNumbers) {
        if (this.threeFirstAreNumbers) {
          message += 'TO HAVE the three first parts to contain numbers.'
        } else {
          message += 'NOT to have numbers at the three first parts.'
        }
        throw new Error(message)
      }
    }
    if (!Array.prototype.includes.call(this.lines, line)) {
      this.lines.push(line)
    }
  }

  getLines(): string[] {
    return [...this.lines]
  }

  getYears(): number[] {
    return extractYears(this.lines)
  }

  getMontsFor(year: number): number[] {
    return extractMonths(this.lines)(year)
  }

  findByDate(year: number = 0, month: number = 0): string[] {
    if (this.threeFirstAreNumbers === false) {
      const message = `This method can only be used when the collection has lines with first part containing a date format (e.g. 2019-05-21-Foo)`
      throw new Error(message)
    }
    return createFindByDate(this.lines)(year, month)
  }
}
