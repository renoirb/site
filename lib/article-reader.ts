import yaml from 'js-yaml'

export const extractFrontMatter = (textContents: string): [any, number] => {
  const lines: string[] = textContents.split('\n')
  let firstContentLine: number = 0
  let hasFoundSecond = false
  const fm: string[] = lines.filter((line, index) => {
    if (/^---$/.test(line) && index > 1) {
      hasFoundSecond = true
      firstContentLine = index + 1
    }
    return hasFoundSecond === false
  })

  const extracted = yaml.safeLoad(fm.join('\n'))

  return [extracted, firstContentLine]
}

export const extractMarkdown = (
  textContents: string,
  firstLine: number
): string => {
  const lines: string[] = textContents.split('\n')
  const out: string[] = lines.filter((line, index) => {
    return index > firstLine
  })

  return out.join('\n')
}
