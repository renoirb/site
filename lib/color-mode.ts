import { Context } from '@nuxt/types'

export const getColorModeClassName = (ctx: Context): string => {
  const colorMode = ctx.app.$colorMode.value
  const colorModeClassName = `${colorMode || 'light'}-mode`
  console.log('colorModeClassName', { colorModeClassName, colorMode })
  return colorModeClassName
}

export const getColorModeFromElement = ($el: Element): string => {
  if ('ownerDocument' in $el === false) {
    const message = `Invalid argument, we expected an Element with an ownerDocument property, but received: ${$el}`
    throw new TypeError(message)
  }
  const doc = $el.ownerDocument
  const [colorMode = 'light-mode'] = [...doc.documentElement.classList].filter(
    (c) => c.endsWith('-mode'),
  )
  return colorMode
}
