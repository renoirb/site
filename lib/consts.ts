import { trimText } from './runtime'

/**
 * This file is imported by any part of the project build
 * so if it's a build too, you can safely import that file.
 * Also, make sure not to import anything.
 */

/**
 * RegExp to ensure baseUrl has no trailing slash
 */
export const RE_BASE_URL = /^https?:\/\/[a-z0-9\-_.]+(?::\d+)?$/i

/**
 * RegExp to check whether it is a blog index or reading one
 */
export const RE_FULL_PATH_BLOG_INDEX = /^\/blog(?:\/tag)?(?:\/\d{4})?(?:\/\d{2})?$/i

/**
 * Store static asset in separate GitHub gh-pages repo, here is the public origin.
 */
export const PUBLIC_SITE_ASSETS_ORIGIN = 'https://renoirb.github.io/site-assets'

/**
 * Fallback image
 */
export const FALLBACK_BLANK_IMAGE = trimText`
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQMA
AAAl21bKAAAAA1BMVEUhR2EbeJkBAAAAAXRSTlPM0jRW/QAAAApJREFUeJ
xjYgAAAAYAAzY3fKgAAAAASUVORK5CYII=
`

/**
 * Where this site will be hosted as
 */
export const PUBLIC_BASE_URL = 'https://renoirb.com'

export const FALLBACK_LOCALE = 'en-CA'

export const FALLBACK_LANG_CODE = FALLBACK_LOCALE.split('-')[0]

export const FALLBACK_TODAY_DATE = new Date().toISOString()

/**
 * When is something published before might be considered
 * out of date?
 */
export const YEAR_CONSIDERED_OLD = 2017

export const PRODUCTION_BASE_PATH = '/site'

export const IS_CI =
  ('IS_CI' in process.env && typeof process.env.IS_CI === 'string') ||
  'GITHUB_ACTION' in process.env
