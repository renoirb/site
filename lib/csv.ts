import { SlugInterface, ArticleFactory, ArticleType } from './models'
import { NuxtRouteInterface } from './runtime'

export const csvToSlugCollection = (type: ArticleType) => {
  const factory = new ArticleFactory(type)
  return (recv: string[]): SlugInterface[] => {
    const articles: SlugInterface[] = []
    recv.forEach(slug => {
      const article = factory.create(slug)
      articles.push(article)
    })

    return articles
  }
}

export function routesCollection(
  slugs: SlugInterface[] = []
): NuxtRouteInterface[] {
  const routes: NuxtRouteInterface[] = []

  return routes
}
