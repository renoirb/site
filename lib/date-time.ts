export type FilterTypeFindByDate = (year: number, month?: number) => string[]

export const extractDateTuple = (line: string): number[] | false => {
  const pipeline = String(line)
    .split('-')
    .slice(0, 3)
    .map(n => parseInt(n, 10))

  const threeFirstAreNotNumbers = pipeline
    .map(i => Number.isInteger(i))
    .includes(false)
  const threeFirstAreNumbers = threeFirstAreNotNumbers === false
  if (threeFirstAreNumbers && pipeline.length === 3) {
    return pipeline
  } else {
    return false
  }
}

export const extractYears = (items: string[]): number[] => {
  const years = new Set<number>()
  for (const line of items) {
    const dateTuple = extractDateTuple(line)
    years.add(dateTuple[0])
  }

  return [...years].sort()
}

export const extractMonths = (items: string[]) => {
  const list = createFindByDate(items)
  return (year: number): number[] => {
    const months = new Set<number>()
    const subList = list(year)
    for (const line of subList) {
      const dateTuple = extractDateTuple(line)
      months.add(dateTuple[1])
    }

    return [...months].sort()
  }
}

export const createFindByDate = (items: string[]): FilterTypeFindByDate => (
  year: number = 0,
  month: number = 0
): string[] => {
  const lines = [...items].filter(line => {
    const dateTuple = extractDateTuple(line)
    if (dateTuple[0] === year) {
      if (month > 0) {
        if (dateTuple[1] === month) {
          return true
        } else {
          return false
        }
      }
      return true
    }
    return false
  })
  return lines
}
