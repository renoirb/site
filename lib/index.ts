export * from './consts'
export * from './model'
export * from './runtime'
