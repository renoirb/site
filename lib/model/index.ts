export * from './content'
export * from './labeler'
