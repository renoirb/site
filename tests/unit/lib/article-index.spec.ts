import { ArticleIndex } from '~/lib/article-index'

const fixture: ReadonlyArray<string> = [
  '2013-02-21-conference-comment-evaluer-la-qualite-dun-site-web-selon-les-techniques-dintegration-web-dactualite.md',
  '2012-11-22-why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client.md',
  '2012-10-27-my-first-introduction-to-the-hypermedia.md',
  '2007-06-26-une-boite-qui-traine.md',
  '2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt.md',
]

const fixtureWithPaths: ReadonlyArray<string> = [
  'anything-here-goes/2013/02/21/conference-comment-evaluer-la-qualite-dun-site-web-selon-les-techniques-dintegration-web-dactualite',
  'anything-here-goes/2012/11/22/why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client',
  'anything-here-goes/2012/10/27/my-first-introduction-to-the-hypermedia',
  'anything-here-goes/2007/06/26/une-boite-qui-traine',
  'anything-here-goes/2009/09/14/je-prefere-le-pingouin-design-cool-de-tee-shirt',
]

describe('ArticleIndex', () => {
  describe('constructor', () => {
    test('Happy-Path', () => {
      const path = 'blog/index.csv'
      const subject = new ArticleIndex(path)
      expect(subject).toHaveProperty('indexFile', path)
    })
  })

  describe('add', () => {
    test('Snapshot when hasPaths is false', () => {
      const subject = new ArticleIndex('bogus')
      fixture.forEach(l => subject.add(l))
      expect(subject).toMatchSnapshot()
    })

    test('Snapshot when hasPaths is true', () => {
      const subject = new ArticleIndex('bogus')
      fixtureWithPaths.forEach(l => subject.add(l))
      expect(subject).toMatchSnapshot()
    })

    test('add enforce consistency when hasPaths is false', () => {
      const subject = new ArticleIndex('bogus')
      // add lines with no slash in them
      // e.g. 2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt.md
      fixture.forEach(l => subject.add(l))
      expect(() => {
        subject.add('foo/bar/baz/buzz')
      }).toThrowError()
    })

    test('add enforce consistency when hasPaths is true', () => {
      const subject = new ArticleIndex('bogus')
      // add lines WITH a slash in them
      fixtureWithPaths.forEach(l => subject.add(l))
      expect(() => {
        subject.add('foo-bar-baz-buzz')
      }).toThrowError()
      expect(() => {
        subject.add(
          '2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt'
        )
      }).toThrowError()
    })

    test('add enforce consistency when threeFirstAreNumbers is true', () => {
      const subject = new ArticleIndex('bogus')
      // add lines with no slash in them
      // e.g. 2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt.md
      fixture.forEach(l => subject.add(l))
      expect(() => {
        subject.add('foo-bar-baz-buzz')
      }).toThrowError()
      // And we cannot mix with using slashes too
      expect(() => {
        subject.add('foo/bar/buzz')
      }).toThrowError()
    })

    test('add enforce consistency when threeFirstAreNumbers is false', () => {
      const subject = new ArticleIndex('bogus')
      subject.add('foo-bar')
      subject.add('baz-buzz')
      expect(() => {
        subject.add('2009-09-14-foo-bar-baz-buzz')
      }).toThrowError()
      // And we cannot mix with using slashes too
      expect(() => {
        subject.add('foo/bar/buzz')
      }).toThrowError()
      // findByDate should not work
      expect(() => {
        subject.findByDate(2019)
      }).toThrowError()
    })
  })

  describe('getLines', () => {
    test('Snapshot when hasPaths is false', () => {
      const subject = new ArticleIndex('bogus')
      fixture.forEach(l => subject.add(l))
      expect(subject.getLines()).toMatchSnapshot()
    })
  })

  describe('findByDate', () => {
    test('Happy-Path', () => {
      const subject = new ArticleIndex('bogus')
      fixture.forEach(l => subject.add(l))
      let found = subject.findByDate(2007)
      expect(found).toEqual(
        expect.arrayContaining(['2007-06-26-une-boite-qui-traine.md'])
      )
      expect(found).toHaveLength(1)
      found = subject.findByDate(2012, 11)
      expect(found).toEqual(
        expect.arrayContaining([
          '2012-11-22-why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client.md',
        ])
      )
      expect(found).toHaveLength(1)
      expect(subject.getLines()).toHaveLength(fixture.length)
    })
  })
})

describe('getYears', () => {
  let subject: ArticleIndex = new ArticleIndex()
  subject.setLines([...fixture])
  let years: number[] = subject.getYears()

  for (const year of years) {
    test(`Looping for findByDate(${year})`, () => {
      const items = subject.findByDate(year)
      expect(items).toMatchSnapshot()
    })
  }

  test('Happy-Path', () => {
    subject = new ArticleIndex()
    subject.setLines([...fixture])
    years = subject.getYears()
    expect(years).toMatchObject([2007, 2009, 2012, 2013])
    expect(years).toMatchSnapshot()
  })
})
