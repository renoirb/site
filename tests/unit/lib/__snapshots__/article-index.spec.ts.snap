// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleIndex add Snapshot when hasPaths is false 1`] = `
ArticleIndex {
  "hasPaths": false,
  "indexFile": "bogus",
  "lines": Array [
    "2013-02-21-conference-comment-evaluer-la-qualite-dun-site-web-selon-les-techniques-dintegration-web-dactualite.md",
    "2012-11-22-why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client.md",
    "2012-10-27-my-first-introduction-to-the-hypermedia.md",
    "2007-06-26-une-boite-qui-traine.md",
    "2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt.md",
  ],
  "threeFirstAreNumbers": true,
}
`;

exports[`ArticleIndex add Snapshot when hasPaths is true 1`] = `
ArticleIndex {
  "hasPaths": true,
  "indexFile": "bogus",
  "lines": Array [
    "anything-here-goes/2013/02/21/conference-comment-evaluer-la-qualite-dun-site-web-selon-les-techniques-dintegration-web-dactualite",
    "anything-here-goes/2012/11/22/why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client",
    "anything-here-goes/2012/10/27/my-first-introduction-to-the-hypermedia",
    "anything-here-goes/2007/06/26/une-boite-qui-traine",
    "anything-here-goes/2009/09/14/je-prefere-le-pingouin-design-cool-de-tee-shirt",
  ],
  "threeFirstAreNumbers": false,
}
`;

exports[`ArticleIndex getLines Snapshot when hasPaths is false 1`] = `
Array [
  "2013-02-21-conference-comment-evaluer-la-qualite-dun-site-web-selon-les-techniques-dintegration-web-dactualite.md",
  "2012-11-22-why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client.md",
  "2012-10-27-my-first-introduction-to-the-hypermedia.md",
  "2007-06-26-une-boite-qui-traine.md",
  "2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt.md",
]
`;

exports[`getYears Happy-Path 1`] = `
Array [
  2007,
  2009,
  2012,
  2013,
]
`;

exports[`getYears Looping for findByDate(2007) 1`] = `
Array [
  "2007-06-26-une-boite-qui-traine.md",
]
`;

exports[`getYears Looping for findByDate(2009) 1`] = `
Array [
  "2009-09-14-je-prefere-le-pingouin-design-cool-de-tee-shirt.md",
]
`;

exports[`getYears Looping for findByDate(2012) 1`] = `
Array [
  "2012-11-22-why-would-i-not-use-my-own-framework-and-how-i-sell-usage-of-symfony2-and-other-current-php-5-3-goodies-to-my-client.md",
  "2012-10-27-my-first-introduction-to-the-hypermedia.md",
]
`;

exports[`getYears Looping for findByDate(2013) 1`] = `
Array [
  "2013-02-21-conference-comment-evaluer-la-qualite-dun-site-web-selon-les-techniques-dintegration-web-dactualite.md",
]
`;
